source/libraries/CompileSolidity.ts

```diff
@@ -0,0 +1,88 @@
+        try {
+            // Compile all contracts in the specified input directory
+            const inputJson: CompilerInput = await this.generateCompilerInput();
+            if (inputJson.error) {
+                throw new Error("The following error was encountered while attempting to generate compiler input:\n\n" + inputJson.error);
+            }
+            const compilerOutput: any = compileStandardWrapper(JSON.stringify(inputJson), this.readCallback);
+            const compileOutputJson = JSON.parse(compilerOutput);
+            if (compileOutputJson.errors) {
+                let errors = "";
+                for (let error of compileOutputJson.errors) {
+                    errors += error.formattedMessage + "\n";
+                }
+                throw new Error("The following errors/warnings were returned by solc:\n\n" + errors);
+            }
+
+            // Create output directory (if it doesn't exist)
+            mkdirp(this.contractOutputDirectoryPath);
+
+            // Output contract data to single file
+            const contractOutputFilePath = this.contractOutputDirectoryPath + "/" + this.contractOutputFileName;
+            let wstream: any = fs.createWriteStream(contractOutputFilePath);
+            for (let contract in compileOutputJson.contracts) {
+                wstream.write(JSON.stringify(compileOutputJson.contracts[contract]));
+            }
+
+            return { output: "Contracts in " + this.contractInputDirectoryPath + " were successfully compiled by solc and saved to " + contractOutputFilePath};
+        } catch (error) {
+            return { error: error.message };
+        }
+    }
+
+    public readCallback(path: string): { contents?: string, error?: string } {
+        try {
+            const result = fs.readFileSync(path, 'utf8');
+            return { contents: result };
+        } catch (error) {
+            console.log(error);
+            return { error: error.message };
+        }
+    }
+
+    private ignoreFile(file: string, stats: fs.Stats): boolean {
+        return stats.isFile() && path.extname(file) != ".sol";
+    }
+
+    private async generateCompilerInput(): Promise<CompilerInput> {
+        let inputJson: CompilerInput = {
+            "language": "Solidity",
+            "sources": {}
+        };
+        try {
+            let contractInputDirectoryPath = this.contractInputDirectoryPath;
+            if (contractInputDirectoryPath.lastIndexOf(path.sep) != contractInputDirectoryPath.length) {
+                contractInputDirectoryPath += path.sep;
+            }
+
+            const files: any = await recursiveReadDir(this.contractInputDirectoryPath, [this.ignoreFile]);
+            for (let index in files) {
+                inputJson.sources[files[index].replace(contractInputDirectoryPath, "")] = {"urls": [files[index]]};
+            }
+        } catch (error) {
+            inputJson.error = error.message;
+        }
```

Comment:
Same comment here as above, by returning the error rather than letting the error bubble, the caller can't easily handle failure vs success separately.  In the case of the caller of this method, they won't want to call `compile` if this function fails, and they probably will want to bail completely.

My general strategy for error handling is to let thrown errors bubble as high up as necessary until they can be cleanly handled.  Often this means letting the errors bubble all the way up the stack and crash the program, because there is no obvious mechanism for recovery.  This is OK because Node will show the error to the user and in the case of scripts like this, a developer will generally be the one running them and can take action to remedy the error and try again.

Reply:
Sounds good.  That's how I've typically done error handling, but I saw you were returning it within a JSON object in `readCallback()`.  I guess this is what solc is expecting from the callback function, and there's no other way to do it?

Reply:
Exactly correct.  The only reason we are returning that `{contents;errors}` object from the `readCallback` is because that is what `solc` needs.  In all other cases I'm a fan of throwing.  Of course, follow whatever pattern you are most comfortable with, even if I include some different pattern in a code snippet or comment, you will be the owner of this project so it should be written in your image not mine.

In general, I'll often give feedback on a bunch of things but in most cases you can consider it just "friendly feedback" and not a requirement.  If something is a hard requirement I'll usually flag the PR as needs changes and mention which points of feedback in particular need to be addressed.  This is usually limited to things that are simply wrong/incorrect.